### PR TITLE
[PLA-2033] Changes for v2.0.0 on marketplace

### PR DESCRIPTION
**Pull request by leonardocustodio**
### **PR Type**
Enhancement, Tests

___

### **Description**
- Introduced a new listing type `OFFER` in the `ListingType` enum and updated related logic.
- Refactored `CreateListingMutation` to use `listingData` instead of `auctionData`, supporting multiple listing types.
- Added new GraphQL input types `ListingDataInputType` and `OfferParamsInputType` to handle diverse listing data.
- Updated database schema with new fields `offer_expiration` and `counter_offer_count` for offers.
- Enhanced tests to cover new listing types and parameters.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>13 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>MarketplaceListingFactory.php</strong><dd><code>Update factory for new listing types and fields</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

database/factories/MarketplaceListingFactory.php

<li>Renamed <code>start_block</code> and <code>end_block</code> to <code>auction_start_block</code> and <br><code>auction_end_block</code>.<br> <li> Added new fields <code>offer_expiration</code> and <code>counter_offer_count</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-marketplace/pull/66/files#diff-d47daae2a29efe1485d710886ec145a597e78d3d93faf9277ed051e1d22c4810">+4/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>2024_10_11_143012_new_listing_type_to_marketplace_listings_table.php</strong><dd><code>Add migration for new listing types and fields</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

database/migrations/2024_10_11_143012_new_listing_type_to_marketplace_listings_table.php

<li>Added migration to rename columns and add new fields for offers.<br> <li> Introduced <code>offer_expiration</code> and <code>counter_offer_count</code> columns.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-marketplace/pull/66/files#diff-9fa1daf0841e3687d98c542ef49acbcef48905a84b7835f97ec8f993be951101">+32/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ListingType.php</strong><dd><code>Add OFFER case to ListingType enum</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/Enums/ListingType.php

- Added a new case `OFFER` to the `ListingType` enum.

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-marketplace/pull/66/files#diff-a65c40fb32fa7b2133ec9f18c76f73fa23854065d53a562ba75d0527aa214590">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>CreateListingMutation.php</strong><dd><code>Refactor CreateListingMutation for multiple listing types</code></dd></summary>
<hr>

src/GraphQL/Mutations/CreateListingMutation.php

<li>Replaced <code>auctionData</code> with <code>listingData</code> to support multiple listing <br>types.<br> <li> Updated validation rules to accommodate new listing types.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-marketplace/pull/66/files#diff-a057c02713a16ae9972ab4e94e73c64d96fe43e6e382c90ca5d1f7c73f64d76e">+60/-47</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ListingDataInputType.php</strong><dd><code>Add ListingDataInputType for diverse listing inputs</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/GraphQL/Types/Input/ListingDataInputType.php

<li>Introduced <code>ListingDataInputType</code> for handling various listing data <br>inputs.<br> <li> Defined fields for <code>type</code>, <code>auctionParams</code>, and <code>offerParams</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-marketplace/pull/66/files#diff-ebcb0cc9c6ebf891d414730b535ed5fb9f99a94cf317b8957e274cb740a7a6f1">+41/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>OfferParamsInputType.php</strong><dd><code>Introduce OfferParamsInputType for offer parameters</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/GraphQL/Types/Input/OfferParamsInputType.php

<li>Created <code>OfferParamsInputType</code> to handle offer-specific parameters.<br> <li> Defined <code>expiration</code> field for offers.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-marketplace/pull/66/files#diff-3ab1c264a6f4e2fd37cecb95fc2aad337c3e4e35a0cd2a6eebfba8482b92102c">+33/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>OfferDataType.php</strong><dd><code>Add OfferDataType for GraphQL offer representation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/GraphQL/Types/OfferDataType.php

<li>Added <code>OfferDataType</code> to represent offer data in GraphQL.<br> <li> Included fields for <code>type</code> and <code>expiration</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-marketplace/pull/66/files#diff-02911b89bdeba40d58982cfc2ac3af354ddd32b93acceac7052f49235c6f3e59">+37/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>OfferStateType.php</strong><dd><code>Add OfferStateType for managing offer states</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/GraphQL/Types/OfferStateType.php

<li>Introduced <code>OfferStateType</code> for managing offer states in GraphQL.<br> <li> Included <code>counterOfferCount</code> field.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-marketplace/pull/66/files#diff-cfe0e2d6aec9aadd20306e6ac47a16edbca622958756e961c3877d0c27d4e446">+37/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ListingDataUnion.php</strong><dd><code>Extend ListingDataUnion with OfferData</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/GraphQL/Unions/ListingDataUnion.php

- Added `OfferData` to the `ListingDataUnion`.

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-marketplace/pull/66/files#diff-a7c13b7175578005ddff81b4a7a676eb8b59a39a31576eae8210190a787c1599">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ListingStateUnion.php</strong><dd><code>Extend ListingStateUnion with OfferState</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/GraphQL/Unions/ListingStateUnion.php

- Added `OfferState` to the `ListingStateUnion`.

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-marketplace/pull/66/files#diff-fa4f6930665018c1fa5a1f05b4abe9ed37e30e3d0e3f6c0e8750fe64824d96f4">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>MarketplaceServiceProvider.php</strong><dd><code>Register migration for new listing types</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/MarketplaceServiceProvider.php

<li>Registered new migration for listing types in the service provider.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-marketplace/pull/66/files#diff-012b810b654b686089fde07eb6428b56a12b68793aea8b5a4f2bc9e83a369700">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ListingDataParams.php</strong><dd><code>Add ListingDataParams for encapsulating listing data</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/Models/Substrate/ListingDataParams.php

<li>Created <code>ListingDataParams</code> to encapsulate listing data parameters.<br> <li> Supported different listing types with specific parameters.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-marketplace/pull/66/files#diff-823b7c655d2bde8d29e61929e7517715d2a16522ed0716739ae4975af83244f2">+33/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>OfferDataParams.php</strong><dd><code>Add OfferDataParams for offer-specific data</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/Models/Substrate/OfferDataParams.php

- Introduced `OfferDataParams` for handling offer-specific data.

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-marketplace/pull/66/files#diff-3d23070be50964094ca1b59575e922a016ff13851f76475f3889021aa6510619">+21/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>CreateListingTest.php</strong><dd><code>Update CreateListing tests for new listing types</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/Feature/GraphQL/Mutations/CreateListingTest.php

<li>Updated tests to use <code>listingData</code> instead of <code>auctionData</code>.<br> <li> Added tests for new listing types and parameters.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-marketplace/pull/66/files#diff-2cac61607acdb6da8186d84f08a33b881518f9b3f4c6132256fb13e432d33d63">+49/-33</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>CreateListing.graphql</strong><dd><code>Update CreateListing mutation to use listingData</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/Feature/GraphQL/Resources/CreateListing.graphql

- Replaced `auctionData` with `listingData` in the mutation.

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-marketplace/pull/66/files#diff-0b24664abac03e3d4d70fe9a49b6e833df3db38aeeed5d3d8394dda6631d632b">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: src/Models/Substrate/OfferDataParams.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    public function __construct(
+        public ?int $expiration = null,
+    ) {}
+
+    /**
+     * Convert the object to encodable formatted array.
+     */
+    public function toEncodable(): array
+    {
+        return ['startBlock' => $this->startBlock, 'endBlock' => $this->endBlock];
```
Comment (v16Studios) [inline src/Models/Substrate/OfferDataParams.php:19]: Just checking, is this the right encodable params for the offer type?  Or should we be encoding the expiration?

File: src/Services/Processor/Substrate/Events/Implementations/Marketplace/ListingCreated.php [hunk 2]
```diff
@@ -7,7 +7,7 @@
 use Enjin\Platform\Marketplace\Enums\ListingState;
 use Enjin\Platform\Marketplace\Enums\ListingType;
 use Enjin\Platform\Marketplace\Events\Substrate\Marketplace\ListingCreated as ListingCreatedEvent;
-use Enjin\Platform\Marketplace\Models\MarketplaceListing;
+use Enjin\Platform\Marketplace\Models\Laravel\MarketplaceListing;
```
Comment (v16Studios) [inline src/Services/Processor/Substrate/Events/Implementations/Marketplace/ListingCreated.php:10]:
```suggestion
use Enjin\Platform\Marketplace\Models\MarketplaceListing;
```
Can we use the Model Resolver version here?

File: tests/Feature/GraphQL/Mutations/CreateListingTest.php [hunk 3]
```diff
@@ -75,9 +73,11 @@ public function test_it_can_skip_validation(): void
...
                 'price' => fake()->numberBetween(1, 1000),
                 'salt' => fake()->text(10),
-                'auctionData' => [
-                    'startBlock' => fake()->numberBetween(1011, 5000),
-                    'endBlock' => fake()->numberBetween(5001, 10000),
+                'listingData' => [
+                    'type' => ListingType::AUCTION->name,
+                    'auctionParams' => ['startBlock' => fake()->numberBetween(1011, 5000),
+                        'endBlock' => fake()->numberBetween(5001, 10000),
+                    ],
```
Comment (v16Studios) [inline tests/Feature/GraphQL/Mutations/CreateListingTest.php:80]:
```suggestion
                'listingData' => [
                    'type' => ListingType::AUCTION->name,
                    'auctionParams' => [
                        'startBlock' => fake()->numberBetween(1011, 5000),
                        'endBlock' => fake()->numberBetween(5001, 10000),
                    ],
```
Just some formatting.

File: tests/Feature/GraphQL/Mutations/CreateListingTest.php [hunk 4]
```diff
@@ -374,79 +374,95 @@ public function test_it_will_fail_with_invalid_parameter_auction_data(): void
...
             $response['error']
         );
 
         $response = $this->graphql(
             $this->method,
-            array_merge($data, ['auctionData' => [
-                'startBlock' => 1012,
-                'endBlock' => 1011,
+            array_merge($data, ['listingData' => [
+                'type' => ListingType::AUCTION->name,
```
Comment (v16Studios) [inline tests/Feature/GraphQL/Mutations/CreateListingTest.php:453]: Please could we add tests for the Offer type as well?

**Review by v16Studios — DISMISSED** [review]

File: src/Services/Processor/Substrate/Events/Implementations/Marketplace/ListingCreated.php (hunk 2 above)
Comment (leonardocustodio) [reply src/Services/Processor/Substrate/Events/Implementations/Marketplace/ListingCreated.php:10]: Doesn't work, will give a type error. Would have to use just Model but that takes type hint away

File: src/Models/Substrate/OfferDataParams.php (hunk 1 above)
Comment (leonardocustodio) [reply src/Models/Substrate/OfferDataParams.php:19]: Thanks, I will check later today.

File: tests/Feature/GraphQL/Mutations/CreateListingTest.php (hunk 4 above)
Comment (leonardocustodio) [reply tests/Feature/GraphQL/Mutations/CreateListingTest.php:454]: Sure 😀

**Review by enjinabner — APPROVED** [review]

**Comment by enjinabner** [comment]
Tests are failing

File: src/Services/Processor/Substrate/Events/Implementations/Marketplace/ListingCreated.php (hunk 2 above)
Comment (v16Studios) [reply src/Services/Processor/Substrate/Events/Implementations/Marketplace/ListingCreated.php:10]: That's fine, we use Model elsewhere for this reason too so should try to keep it consistent 👍🏻

**Pull request merged**